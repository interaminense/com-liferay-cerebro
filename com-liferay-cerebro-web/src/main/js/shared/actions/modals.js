export const actionTypes = {
	CLOSE_ALL_MODALS: 'CLOSE_ALL_MODALS',
	CLOSE_MODAL: 'CLOSE_MODAL',
	OPEN_MODAL: 'OPEN_MODAL'
};

export const modalTypes = {
	CREATE_MAPPING_MODAL: 'CREATE_MAPPING_MODAL',
	CSV_PREVIEW_MODAL: 'CSV_PREVIEW_MODAL',
	EDIT_CARD_MODAL: 'EDIT_CARD_MODAL',
	LOADING_MODAL: 'LOADING_MODAL',
	MATCH_FIELDS_MODAL: 'MATCH_FIELDS_MODAL',
	SELECT_ITEMS_MODAL: 'SELECT_ITEMS_MODAL',
	TEST: 'TEST'
};

export function close() {
	return {
		type: actionTypes.CLOSE_MODAL
	};
}

export function closeAll() {
	return {
		type: actionTypes.CLOSE_ALL_MODALS
	};
}

export function open(type, props = {}, options = {}) {
	const {closeOnBlur = true} = options;

	return {
		payload: {
			closeOnBlur,
			props,
			type
		},
		type: actionTypes.OPEN_MODAL
	};
}