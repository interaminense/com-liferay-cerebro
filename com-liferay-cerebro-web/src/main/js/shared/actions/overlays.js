export const actionTypes = {
	HIDE_OVERLAY: 'HIDE_OVERLAY',
	SHOW_OVERLAY: 'SHOW_OVERLAY'
};

export function hide(id) {
	return {
		payload: {
			id
		},
		type: actionTypes.HIDE_OVERLAY
	};
}

export function show(options) {
	const {
		alignment,
		containerClass,
		context,
		forceAlignment,
		id,
		offset,
		onMouseEnter,
		onMouseLeave,
		onOutsideClick,
		parentId
	} = options;

	return {
		payload: {
			alignment,
			containerClass,
			context,
			forceAlignment,
			id,
			offset,
			onMouseEnter,
			onMouseLeave,
			onOutsideClick,
			parentId
		},
		type: actionTypes.SHOW_OVERLAY
	};
}