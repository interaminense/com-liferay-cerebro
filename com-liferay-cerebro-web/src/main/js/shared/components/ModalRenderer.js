import autobind from 'autobind-decorator';
import Component, {Config} from 'metal-jsx';
import dom from 'metal-dom';
import {connect} from 'metal-redux';
import {List} from 'immutable';

import CreateMappingModal from './CreateMappingModal';
import LoadingModal from './LoadingModal';
import SelectItemsModal from './SelectItemsModal';
import TestModal from './TestModal';
import {close, modalTypes} from '../actions/modals';

const BODY_CLASSNAME = 'modal-open';

const COMPONENT_MAP = {
	[modalTypes.CREATE_MAPPING_MODAL]: CreateMappingModal,
	[modalTypes.LOADING_MODAL]: LoadingModal,
	[modalTypes.SELECT_ITEMS_MODAL]: SelectItemsModal,
	[modalTypes.TEST]: TestModal
};

function toggleBodyModalOpen(open = true) {
	if (open) {
		document.body.classList.add(BODY_CLASSNAME);
	} else {
		document.body.classList.remove(BODY_CLASSNAME);
	}
}

class ModalRenderer extends Component {
	rendered() {
		toggleBodyModalOpen(!!this.getCurrentModal());
	}

	disposed() {
		toggleBodyModalOpen(false);
	}

	getCurrentModal() {
		return this.props.modalsIList.last();
	}

	@autobind
	handleClickOutside(event) {
		const currentModalIMap = this.getCurrentModal();

		if (
			currentModalIMap.get('closeOnBlur', true) &&
			dom.match(event.target, '.modal-container')
		) {
			this.props.close();
		}
	}

	render() {
		return (
			<div class="modal-renderer-root">
				{this.props.modalsIList
					.map((modalIMap, i) => {
						const ModalComponent =
							COMPONENT_MAP[modalIMap.get('type')];

						return (
							<div
								class="modal-container d-block fade modal show"
								key={i}
								onClick={this.handleClickOutside}
								role="dialog"
							>
								<ModalComponent
									{...modalIMap.get('props').toJS()}
								/>
							</div>
						);
					})
					.toJS()}

				{!!this.getCurrentModal() && (
					<div class="modal-backdrop fade show" />
				)}
			</div>
		);
	}
}

const STORE = {
	close: Config.func().required(),
	modalsIList: Config.instanceOf(List).required()
};

ModalRenderer.PROPS = {
	...STORE
};

export default connect(
	state => ({
		modalsIList: state.get('modals', List())
	}),
	{close}
)(ModalRenderer);