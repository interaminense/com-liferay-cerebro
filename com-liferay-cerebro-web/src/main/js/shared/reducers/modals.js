import {createReducer} from 'redux-toolbox';
import {fromJS, List} from 'immutable';

import {actionTypes} from '../actions/modals';

export default createReducer(List(), {
	[actionTypes.CLOSE_ALL_MODALS](state) {
		return state.clear();
	},

	[actionTypes.CLOSE_MODAL](state) {
		return state.pop();
	},

	[actionTypes.OPEN_MODAL](state, action) {
		return state.push(fromJS(action.payload));
	}
});