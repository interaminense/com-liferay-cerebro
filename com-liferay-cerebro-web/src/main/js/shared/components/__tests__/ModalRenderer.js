jest.mock('../../actions/modals');

import {fromJS} from 'immutable';
import dom from 'metal-dom';

import mockStore from 'test/mock-store';
import ModalRenderer from '../ModalRenderer';
import * as modalActions from '../../actions/modals';

const {modalTypes} = modalActions;

describe('ModalRenderer', () => {
	let component;

	afterEach(() => {
		if (component) {
			component.dispose();
		}

		jest.clearAllMocks();
	});

	it('should render', () => {
		const store = mockStore(
			fromJS({
				modals: []
			})
		);

		component = new ModalRenderer({store});

		expect(component).toMatchSnapshot();
	});

	it('should render test modal', () => {
		const store = mockStore(
			fromJS({
				modals: [
					{
						props: {},
						type: modalTypes.TEST
					}
				]
			})
		);

		component = new ModalRenderer({store});

		expect(component).toMatchSnapshot();
	});

	it('should add .modal-open to body', () => {
		const store = mockStore(
			fromJS({
				modals: [
					{
						props: {},
						type: modalTypes.TEST
					}
				]
			})
		);

		expect(document.body.classList.contains('modal-open')).toBe(false);

		component = new ModalRenderer({store});

		expect(document.body.classList.contains('modal-open')).toBe(true);
	});

	it('should render multiple modals', () => {
		const store = mockStore(
			fromJS({
				modals: [
					{
						props: {},
						type: modalTypes.TEST
					},
					{
						props: {},
						type: modalTypes.TEST
					}
				]
			})
		);

		component = new ModalRenderer({store});

		expect(component).toMatchSnapshot();
	});

	it('should pass props to modal', () => {
		const store = mockStore(
			fromJS({
				modals: [
					{
						props: {
							title: 'FooBar'
						},
						type: modalTypes.TEST
					}
				]
			})
		);

		component = new ModalRenderer({store});

		expect(component).toMatchSnapshot("should render with title 'FooBar'");
	});

	it('should close modal on click outside', () => {
		const store = mockStore(
			fromJS({
				modals: [
					{
						props: {
							title: 'FooBar'
						},
						type: modalTypes.TEST
					}
				]
			})
		);

		component = new ModalRenderer({store});

		const element = component.element.querySelector('.modal-container');

		expect(modalActions.close).not.toHaveBeenCalled();

		dom.triggerEvent(element, 'click');

		expect(modalActions.close).toHaveBeenCalled();
	});

	it('should not close modal if closeOnBlur is disabled', () => {
		const store = mockStore(
			fromJS({
				modals: [
					{
						closeOnBlur: false,
						props: {
							title: 'FooBar'
						},
						type: modalTypes.TEST
					}
				]
			})
		);

		component = new ModalRenderer({store});

		const element = component.element.querySelector('.modal-container');

		expect(modalActions.close).not.toHaveBeenCalled();

		dom.triggerEvent(element, 'click');

		expect(modalActions.close).not.toHaveBeenCalled();
	});
});