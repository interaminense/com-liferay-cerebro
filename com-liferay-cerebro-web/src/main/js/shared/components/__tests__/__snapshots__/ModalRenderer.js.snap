// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalRenderer should render 1`] = `<div class="modal-renderer-root"></div>`;

exports[`ModalRenderer should render multiple modals 1`] = `
<div class="modal-renderer-root">
  <div class="modal-container d-block fade modal show" role="dialog">
    <div class="modal-dialog">
      <div class="modal-content">
        <div class="modal-header">
          <h4 class="modal-title">Modal</h4>
          <button class="button-root btn btn-secondary close" type="button">
            <svg class="icon-root lexicon-icon lexicon-icon-times">
              <use xlink:href="/images/lexicon/icons.svg#times"></use>
            </svg>
          </button>
        </div>
        <div class="modal-body inline-scroller">
          <h4>Modal Body</h4>
        </div>
        <div class="modal-footer">
          <button class="button-root btn btn-secondary" type="button">Submit</button>
          <button class="button-root btn btn-secondary" type="button">Cancel</button>
        </div>
      </div>
    </div>
  </div>
  <div class="modal-container d-block fade modal show" role="dialog">
    <div class="modal-dialog">
      <div class="modal-content">
        <div class="modal-header">
          <h4 class="modal-title">Modal</h4>
          <button class="button-root btn btn-secondary close" type="button">
            <svg class="icon-root lexicon-icon lexicon-icon-times">
              <use xlink:href="/images/lexicon/icons.svg#times"></use>
            </svg>
          </button>
        </div>
        <div class="modal-body inline-scroller">
          <h4>Modal Body</h4>
        </div>
        <div class="modal-footer">
          <button class="button-root btn btn-secondary" type="button">Submit</button>
          <button class="button-root btn btn-secondary" type="button">Cancel</button>
        </div>
      </div>
    </div>
  </div>
  <div class="modal-backdrop fade show"></div>
</div>
`;

exports[`ModalRenderer should render test modal 1`] = `
<div class="modal-renderer-root">
  <div class="modal-container d-block fade modal show" role="dialog">
    <div class="modal-dialog">
      <div class="modal-content">
        <div class="modal-header">
          <h4 class="modal-title">Modal</h4>
          <button class="button-root btn btn-secondary close" type="button">
            <svg class="icon-root lexicon-icon lexicon-icon-times">
              <use xlink:href="/images/lexicon/icons.svg#times"></use>
            </svg>
          </button>
        </div>
        <div class="modal-body inline-scroller">
          <h4>Modal Body</h4>
        </div>
        <div class="modal-footer">
          <button class="button-root btn btn-secondary" type="button">Submit</button>
          <button class="button-root btn btn-secondary" type="button">Cancel</button>
        </div>
      </div>
    </div>
  </div>
  <div class="modal-backdrop fade show"></div>
</div>
`;

exports[`should render with title 'FooBar' 1`] = `
<div class="modal-renderer-root">
  <div class="modal-container d-block fade modal show" role="dialog">
    <div class="modal-dialog">
      <div class="modal-content">
        <div class="modal-header">
          <h4 class="modal-title">FooBar</h4>
          <button class="button-root btn btn-secondary close" type="button">
            <svg class="icon-root lexicon-icon lexicon-icon-times">
              <use xlink:href="/images/lexicon/icons.svg#times"></use>
            </svg>
          </button>
        </div>
        <div class="modal-body inline-scroller">
          <h4>Modal Body</h4>
        </div>
        <div class="modal-footer">
          <button class="button-root btn btn-secondary" type="button">Submit</button>
          <button class="button-root btn btn-secondary" type="button">Cancel</button>
        </div>
      </div>
    </div>
  </div>
  <div class="modal-backdrop fade show"></div>
</div>
`;
